neuraxle/checkpoints.py

```diff
@@ -19,185 +19,442 @@
+            data_container: DataContainer,
+            context: ExecutionContext
+    ) -> DataContainer:
+        # TODO: save the context by execution mode AND data container ids / summary
+        context.save_all_unsaved()
+        return data_container
+
+    def should_resume(self, data_container: DataContainer, context: ExecutionContext) -> bool:
+        # TODO: change this when we support multiple execution modes and data container ids / summary
+        return True
+
+
+class BaseDataContainerSaver(ABC):
 
     @abstractmethod
-    def set_checkpoint_path(self, path):
+    def get_checkpoint_file_path(self, checkpoint_path: str, current_id) -> str:
         """
-        Set checkpoint Path
+        Returns the checkpoint file path for a data input id
 
-        :param path: checkpoint path
+        :param checkpoint_path:
+        :param current_id:
+        :return:
         """
         raise NotImplementedError()
 
     @abstractmethod
-    def read_checkpoint(self, data_container: DataContainer) -> DataContainer:
+    def save_current_id_checkpoint(self, checkpoint_path: str, current_id, data_input, expected_output) -> bool:
         """
-        Read checkpoint data to get the data inputs and expected output.
+        Save the given current id, data input, and expected output.
 
-        :param data_container: data inputs to save
-        :return: checkpoint data container
+        :param checkpoint_path:
+        :param current_id:
+        :param data_input:
+        :param expected_output:
+        :return:
         """
         raise NotImplementedError()
 
     @abstractmethod
-    def save_checkpoint(self, data_container: DataContainer) -> DataContainer:
+    def read_current_id_checkpoint(self, checkpoint_path: str, current_id) -> Tuple:
         """
-        Save checkpoint for data inputs and expected outputs so that it can
-        be loaded by the checkpoint pipeline runner on the next pipeline run
+        Read the data inputs, and expected outputs for the given current id.
 
-        :param data_container: data inputs to save
-        :return: saved data container
+        :param checkpoint_path:
+        :param current_id:
+        :return: tuple(current_id, checkpoint_data_input, checkpoint_expected_output)
+        :rtype: tuple(str, Iterable, Iterable)
         """
         raise NotImplementedError()
 
 
-class PickleCheckpointStep(BaseCheckpointStep):
+class PickleDataContainerSaver(BaseDataContainerSaver):
     """
-    Create pickles for a checkpoint step data inputs, and expected_outputs
-    to eventually be able to load them using the checkpoint pipeline runner.
+    PickleDataContainerSaver is used by DataContainerCheckpointer so save,
+    and load the checkpoint files for all of the current ids inside a data container.
     """
+    def __init__(self, execution_mode: ExecutionMode):
+        if execution_mode == ExecutionMode.FIT:
+            self.checkpoint_file_name_suffix = ExecutionMode.FIT.value
```

Comment:
There is a better way to do this, it doesn't solve completely the problem I think. 

Again, here is a requirement: I think we will need to be able to transform data on which we fit_transformed earlier. That is a difficult question thought, because if we're in a KFoldCV random search then we might want to separate different trials completely.

Reply:
I've fixed your other comments @guillaume-chevalier can we call each other to talk about this one (the more crunchy one) so we can find a solution ?